runner.py

```diff
@@ -0,0 +1,707 @@
+
+    def __init__(self, target):
+        if target not in TargetBuilder.TARGETS:
+            raise ValueError(f"invalid target: {target}")
+
+        self.env = os.environ.copy()
+        self.target = target
+        self.fs_path = PRTOS_PROJECT_DIR / f"_fs/{self.target}"
+        self.env['TARGET'] = self.target
+        self.env['CONSOLE'] = 'serial'
+
+    def __str__(self):
+        return self.target
+
+    def fs_mkdir(self, path):
+        if path.anchor == '/':
+            path = path.relative_to('/')
+        abs_path = self.fs_path / path
+        abs_path.mkdir(exist_ok=True)
+
+    def fs_install(self, path, file, mode):
+        if path.anchor == '/':
+            path = path.relative_to('/')
+        abs_path = self.fs_path / path
+        shutil.copy(file, abs_path)
+        abs_path.joinpath(file.name).chmod(mode)
+
+    def prebuild_action(self, syspage=None, **kwargs):
+        pass
+
+    def postbuild_action(self, syspage=None, **kwargs):
+        pass
+
+    def run_command(self, args, handle_error=True):
+        proc = subprocess.Popen(
+            args,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            env=self.env,
+            universal_newlines=True,
+            cwd=PRTOS_PROJECT_DIR
+        )
+
+        out, err = proc.communicate()
+        if proc.returncode != 0 and handle_error:
+            logging.error(f"Command {' '.join(args)} for {self.target} failed!")
+            logging.error(out)
+            logging.error(err)
+            # Fail at build stage should not happen, exit immediately
+            sys.exit(1)
+        elif proc.returncode != 0:
+            logging.debug(f"Command {' '.join(args)} for {self.target} failed!")
+        else:
+            logging.debug(f"Command {' '.join(args)} for {self.target} success!")
+
+        return proc.returncode, out, err
+
+    def build(self, syspage=None):
+        if syspage:
+            syspage_with_defaults = list(set(syspage + TargetBuilder.SYSPAGE[self.target]))
+            self.env['SYSPAGE'] = " ".join(syspage_with_defaults)
+
+        self.prebuild_action(syspage)
+
+        logging.info(f"Building {self.env['TARGET']} with syspage: {self.env['SYSPAGE']}")
+        self.run_command(['./docker-build.sh', 'clean', 'core', 'fs', 'image', 'project'])
+
+        self.postbuild_action(syspage)
+
+
+class IA32Builder(TargetBuilder):
+    def __init__(self):
+        super().__init__('ia32-generic')
+
+    def prebuild_action(self, syspage=None, **kwargs):
+        # We run syspage progs from filesystem, set only default syspage
+        self.env['SYSPAGE'] = " ".join(self.SYSPAGE[self.target])
+
+    def postbuild_action(self, syspage=None, **kwargs):
+        if not syspage:
+            return
+
+        logging.info(f"Installing {' '.join(syspage)} to filesystem for {self.target}")
+        build_dir = PRTOS_PROJECT_DIR / f"_build/{self.target}/prog.stripped"
+        syspage_dir = pathlib.Path("/root/syspage")
```

Comment:
no, install in `/bin`, `/syspage` is for programs which are embedded in kernel binary

Reply:
I changed it to `/bin` however "syspage" handling have to be changed in the future PR. For now it's impossible to include multiple programs (like some driver and user app) into one test case (we do not choose which one we want to execute).

Reply:
I think that the "putting the binaries/files onto the device" and `executing programs` will need to be part of the `DeviceRunner` in the future. For some "big" targets it will be just appending them to the `rootfs` image, but for smaller (like iMX.RT1064) it might involve uploading binaries via USB/serial using `plo` bootloader during the bootup.

Regarding the "test binary execution" part - for now let's assume the "base" image would include all necessary drivers and tools, if the test would need multiple threads/processes it can spawn them internally.